### PR TITLE
Game parser refactor

### PR DESCRIPTION
**Pull request by RoiEXLab**
Follow-up of #2649

Open to suggestions, code might not be perfect.

### CODE REVIEW

File: src/main/java/games/strategy/engine/data/GameParser.java [hunk 1]
```diff
@@ -56,7 +58,7 @@
 
 public class GameParser {
   private static final Class<?>[] SETTER_ARGS = {String.class};
-  private GameData data;
+  private final GameData data = new GameData();
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/GameParser.java]: Although this slightly changes functionality, it's safe to change it, because all callers create fresh instance of GameParser

Comment (ssoloff) [reply src/main/java/games/strategy/engine/data/GameParser.java]: Agree it's safe.  As I mentioned elsewhere, `GameParser` appears to be a _Method Object_, and thus should only ever be "used" once.  It might be useful to update the class Javadocs stating that once one of the public `parseXXX()` methods is called, the object can no longer be used.  Some example code might explain that nicely.

An alternative is to just make it impossible to make that mistake.  You could make the constructor and all instance methods private and just provide public static methods that handle the two use cases.  For example:

```java
public static GameData parse(final String mapName, final InputStream stream, final AtomicReference<String> gameName)
    throws GameParseException, SAXException, EngineVersionException, IllegalArgumentException {
  return new GameParser(mapName).parse(stream, gameName);
}
```
```

File: src/main/java/games/strategy/engine/data/GameParser.java [hunk 2]
```diff
@@ -65,26 +67,35 @@ public GameParser(final String mapName) {
...
   /**
    * Parses a file into a GameData object.
-   *
-   * @param delayParsing Should we only parse the game name, notes, and playerlist? Normally this should be "false",
-   *        except for the game chooser which should use the user set preference.
    */
-  public synchronized GameData parse(final InputStream stream, final AtomicReference<String> gameName,
-      final boolean delayParsing)
+  public synchronized GameData parse(final InputStream stream, final AtomicReference<String> gameName)
       throws GameParseException, SAXException, EngineVersionException, IllegalArgumentException {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/GameParser.java]: Nit: Can probably remove the unchecked `IllegalArgumentException` from the exception specification.  Instead we should document under what conditions it can be thrown.

Actually, it looks like we changed that behavior.  Whereas previously, a `null` stream would trigger an IAE, it now triggers an NPE.  The NPE is probably the correct choice, so no Javadocs should be needed to state that explicitly, as that behavior should be assumed.

**Comment by RoiEXLab** [comment]
@ssoloff I applied your proposed changes (slightly improved), and added a precondition to the constructor.
Merging this now and leaving the rest of your suggestions up to you to implement.
You seem to have a pretty good idea of how this class should look like, but it's a bit hard to follow your individual suggestions that partially imply a complete restrucuring of this class
